### ISSUE TITLE
The problem that the re-projection error is too large

### ISSUE THREAD

**Issue by 88leesy** [cvg/Hierarchical-Localization#84]
Hello, 

I'm a graduate student In South Korea.

I have two issues to write about...

* About Aachen Training Datasets 
First,  I wonder why there is a big difference between the 2D points that projected 3D points onto 2D Image  using the ground truth(GT) camera pose and the given GT 2D points, (that is re-projection error). 

The process is as follows.

 - About Aachen Training datasets
 - reference code : pipeline_Aachen.ipynb (I ran all of the code)
 - For re-projection error check Function : 
    localize_sfm.main() ->  pose_from_cluster()
     -> 1. got (2D points of reference image & matches 2D points)(A) and (correspondence 3D points)(B)
     -> 2 . (2D points)(A') that projected (3D points)(B) onto 2D Image
     -> A - A' compared...  => re-projection error big large!
 
```pyhon

pairs = Path('pairs/aachen/')
sfm_pairs = pairs / 'pairs-db-covis20.txt'  # top 20 most covisible in SIFT model
outputs = Path('outputs/aachen/')  # where everything will be saved
reference_sfm = outputs / 'sfm_superpoint+superglue'  # the SfM model we will build

localize_sfm.main(
    #reference_sfm / 'model',
    reference_sfm,
    dataset / '3D-models/database_intrinsics.txt',
    sfm_pairs,
    outputs / f"{feature_conf['output']}.h5",
    outputs / f"{feature_conf['output']}_{matcher_conf['output']}_{sfm_pairs.stem}.h5",
    results,
    covisibility_clustering=False # not required with SuperPoint+SuperGlue
)

# re-projection error check code:
def pose_from_cluster(qname, qinfo, db_ids, db_images, points3D,
                      feature_file, match_file, thresh, gt_dbid):

    R_gt = db_images[gt_dbid].qvec2rotmat()
    t_gt = db_images[gt_dbid].tvec
    for i in range(len(mkpq)):
        xx = float(mkpq[i][0])
        yy = float(mkpq[i][1])
        cv2.circle(image, (int(xx), int(yy)), 2, (255, 0, 0), -1)

        point3f = np.zeros((4, 1), np.float64)
        point3f[0] = mp3d[i][0]
        point3f[1] = mp3d[i][1]
        point3f[2] = mp3d[i][2]
        point3f[3] = 1
        Rt_gt =np.append(R_gt, t_gt.reshape(3, 1), axis=-1)
        point2f = np.dot(np.dot(camera_metrix, Rt_gt ), point3f)
        proj_x = point2f[0] / point2f[2]
        proj_y = point2f[1] / point2f[2]
        cv2.circle(image, (int(proj_x), int(proj_y)), 2, (0, 255, 0), -1)
        gt_dist_err = math.sqrt( (xx-proj_x)**2 + (yy-proj_y)**2)
        gt_covis_dist_err_list.append(gt_dist_err)
mean_covis_dist_err = np.mean(gt_covis_dist_err_list, axis=0)
print('mean_covis_dist_err:  ', mean_covis_dist_err, end='\n')

```
It is represented graphically as follows :

![Figure_0](https://user-images.githubusercontent.com/56951517/129147772-f2af16e8-1b36-4d43-900a-4698be52518d.png)
opencvPNp colamp index:   db/2738.jpg   Error XYZ (m):   1.4094536013298224   Error Q (degrees):   0.1450112900798425
pycolmap PNP colamp index:   db/2738.jpg   Error XYZ (m):   0.6016059692014329   Error Q (degrees):   0.05977929291767014
mean_covis_dist_err:   152.3036582842934

*mean_covis_dist_error = re-projection error
*red points : GT points
*green  points : projection point
=>Compared to GT and PnP camera pose errors, the reprojection error is relatively high.

Second, Perspective-N-Point error...

- The difference between the camera pose predicted by the GT's 2D-3D correspondence point and the GT's camera pose is too large.

![Figure_1](https://user-images.githubusercontent.com/56951517/129147318-68c55206-8443-434d-a370-041b248fa363.png)
opencvPNp colamp index:   db/2699.jpg   Error XYZ (m):   81.93468187225895   Error Q (degrees):   355.1609927572119
pycolmap PNP colamp index:   db/2699.jpg   Error XYZ (m):   117.19459253913064   Error Q (degrees):   352.9739317473679
mean_covis_dist_err:   824.0163571640451  
*red points : GT points
*green  points : projection point

- I'd appreciate it if you could let me know what I'm doing wrong.
- Thank you.

**Comment by Skydes** [comment]
1. All images of the Aachen dataset have radial distortion, but the code that you show does not apply any undistortion. You can use `pycolmap.world_to_image` to undistort and denormalize the 2D reprojections.
2. It is hard to tell without knowing how the poses are estimated. Again, this needs to take into account the radial distortion.

**Comment by 88leesy** [comment]
Thanks for quick reply.

'''
python

    camera_model, width, height, params = qinfo
    cfg = {
        'model': camera_model,
        'width': width,
        'height': height,
        'params': params,
    }
    ret = pycolmap.absolute_pose_estimation(mkpq, mp3d, cfg, thresh)
    ret['cfg'] = cfg

'''

You know, In practice, pose_from_cluster function(line87~) in localize_sfm.py is applied as above to derive results for query datasets.

I think this is pose estimation applied to camera parameters (including distortion coefficients).

Actually, I've already applied it, also... 

And...  About first question, 

For the first question, I applied the opencv function to obtain the projected 2D points, and the results were similar to the previously obtained projection method.

ex) imgpts, jc = cv2.projectPoints(3D_Points, R_vec, t, camera_matrix, np.array([params[3], 0.0, 0.0, 0.0])),
     params[3]:distortion coefficient

I look forward to your reply, thanks for reading my post.
